### PR TITLE
Handling HTML5 named character references

### PR DESCRIPTION
**Pull request by kennypete**
Refer enhancement issue #19

### CODE REVIEW

**Comment by tpope** [comment]
My first impressions are that this can probably be simplified quite a bit. Move the HTML entity handling out of the loop and operate on `a:char` as a whole. And change the lookup table and #html_entity` to use characters as keys rather than numbers. Use `"\u...."` for the dictionary key literals.

If there are edge cases this wouldn't cover, please do elaborate.

**Comment by kennypete** [comment]
> My first impressions are that this can probably be simplified quite a bit. Move the HTML entity handling out of the loop and operate on `a:char` as a whole. And change the lookup table and `#html_entity` to use characters as keys rather than numbers. Use `"\u...."` for the dictionary key literals.

Yes, partly I'd not wanted to go wider than just adjusting the existing approach, but I'd not really considered optimising it either.  Since that's okay to do, yes, the HTML handling outside the loop works well and characters as the keys to `#html_entity` also works (though just using characters for keys works perfectly, e.g., `'>⃒': 'nvgt'`.

> If there are edge cases this wouldn't cover, please do elaborate.

A couple immediately came to mind: `&NewLine;` and `&fjlig;`.  The former because can't be _selected_ in Vim (as far as I know) and the latter because it is just f and j in succession, so is not "one character".  Even characters that I thought could be iffy like `&it;` and `&lrm;` work fine.  One is peculiar, `&DownBreve;`, which will only display with `ga` when not preceded by a character, otherwise it is combined with whatever precedes it, but I guess that's kind of its point. 


I tested it and could find no characters that would not work (using HTML5chars.txt to the support folder - hope that's okay/delete if not wanted).  I also tried some outside of HTML5 scope, including emoji characters, just in case there was some side effect, but found nothing.  I have re-pushed to my fork.

File: plugin/characterize.vim [hunk 1]
```diff
@@ -30,12 +31,13 @@ function! s:info(char) abort
     for emoji in characterize#emojis(nr)
       let out .= ', '.emoji
     endfor
-    let entity = characterize#html_entity(nr)
-    if !empty(entity)
-      let out .= ', '.entity
-    endif
     call add(outs, out)
   endwhile
+  let entity = characterize#html_entity(matchstr(getline('.')[col('.')-1:-1],'.'))
```
Comment (tpope) [inline plugin/characterize.vim:36]: Use `a:char` here.

Comment (kennypete) [reply plugin/characterize.vim]: Done, tested, no issues. (And, incidentally, my temporarily commented line 17, which I'd not noticed and should have been removed, has been deleted.)

**Comment by tpope** [comment]
> A couple immediately came to mind: `&NewLine;` and `&fjlig;`. The former because can't be _selected_ in Vim (as far as I know) and the latter because it is just f and j in succession, so is not "one character". Even characters that I thought could be iffy like `&it;` and `&lrm;` work fine. One is peculiar, `&DownBreve;`, which will only display with `ga` when not preceded by a character, otherwise it is combined with whatever precedes it, but I guess that's kind of its point.

These sound like limitations of both your original approach and my modified one. We're good to go on that front.

Let's go with `"\u...."` style characters rather than literals. This is closer to the hex number used by the other dictionaries, and avoids weirdness around DownBreve, etc.

**Comment by kennypete** [comment]
> These sound like limitations of both your original approach and my modified one. We're good to go on that front.

Okay, I'd not characterise (pardon the pun) those as limitations, i.e., characters that either cannot be selected insofar as the cursor cannot be on top of the character (line feed) or is solely a composing character (breve).  Good to go is great though. 

> Let's go with `"\u...."` style characters rather than literals. This is closer to the hex number used by the other dictionaries, and avoids weirdness around DownBreve, etc.

I'll do that shortly.  I guess the conversion step from the Unicode character itself to `\\u....` has the advantage of humans being able to _see_ the keys easily (the whitespace variants are not friendly from a key perspective either) and would add immaterial overhead for a one character action, i.e., in terms of taking the Unicode character to the `\\u....` form to look it up.

**Comment by tpope** [comment]
> > Let's go with `"\u...."` style characters rather than literals. This is closer to the hex number used by the other dictionaries, and avoids weirdness around DownBreve, etc.
> 
> I'll do that shortly. I guess the conversion step from the Unicode character itself to `\\u....` has the advantage of humans being able to _see_ the keys easily (the whitespace variants are not friendly from a key perspective either)

Exactly.

> and would add immaterial overhead for a one character action, i.e., in terms of taking the Unicode character to the `\\u....` form to look it up.

The overhead would happen once for all characters, when the autoload file is loaded and the dictionary is created. This dictionary is tiny compared to `s:d`, so I highly doubt it would be noticeable.

**Comment by kennypete** [comment]
Changing to the code points demanded a few collateral changes, e.g., to build the key for the two code point ones requires reading both the character and the composing character to build the key.  

This should be good now I think: I tested on many of the characters in the HTML5entities file, including two code point characters, and also on many of the emoji characters.  All return correct info.

**Comment by tpope** [comment]
The key should look like `"\u1234"`, not `'U+1234'`. If you do that, all the key-building goes away, just look for the character sequence in the dictionary directly.

**Comment by kennypete** [comment]
> The key should look like `"\u1234"`, not `'U+1234'`. If you do that, all the key-building goes away, just look for the character sequence in the dictionary directly.

Sorry, my non-thinking.  It's sorted now.  The two code points ones are straightforward keys, e.g., `"\\u219d\\u0338": '&nrarrw;'` and >FFFF are done, e.g., `"\U1d49c": '&Ascr;'`.

**Comment by kennypete** [comment]
Hi, this is good to merge now?

**Comment by tpope** [comment]
It's still messy. The old code put each code point into the list `outs` and joined them into the final return value. The new version makes one big string, clumsily special casing the first entry, but then shoves that individual string into a one item list anyways.

I tried fixing it locally, but that exposed the conceptual problem that since the HTML entity belongs to _all_ code points, it shouldn't be tacked onto the end with a comma, that implies it belongs to the last code point only. I'm not sure what to do instead.

This is where I left it:

```diff
diff --git i/plugin/characterize.vim w/plugin/characterize.vim
index c112827..f57c486 100644
--- i/plugin/characterize.vim
+++ w/plugin/characterize.vim
@@ -30,13 +30,14 @@ function! s:info(char) abort
     for emoji in characterize#emojis(nr)
       let out .= ', '.emoji
     endfor
-    let entity = characterize#html_entity(nr)
-    if !empty(entity)
-      let out .= ', '.entity
-    endif
     call add(outs, out)
   endwhile
-  return join(outs, ' ')
+  let full = join(outs, ' ')
+  let entities = characterize#html_entities(a:char)
+  if !empty(entities)
+    let full .= ' | ' . entities
+  endif
+  return full
 endfunction
 
 nnoremap <silent><script> <Plug>(characterize) :<C-U>echo <SID>info(matchstr(getline('.')[col('.')-1:-1],'.'))<CR>
```

**Comment by kennypete** [comment]
From what I recall, there appeared no obvious way to: 1) make it elegant, 2) omit outputting an entity for the first of two code points where that first one would be an entity per se but for it being followed by a composing character making it a distinct entity, and 3) output the two code point entity, where applicable.   

I'll look at it again in a few days and see whether the old `while` could be used; perhaps it's better done by testing whether the nr and a:char entities are the same.  Doing that would allow reverting to the old approach (but would require suppressing the entity when it is not the same as the "all"/both code points one).  A final test could then distinguish the "all" entity from its parts, when applicable.  Displaying **|** could visually achieve that, as you have above, though rather than **| \&entity;** it may be better to include the character as well, to be clear that the entity is to to both/all?  E.g., **| <∂̸> \&npart;**

**Comment by kennypete** [comment]
The latest pull request reverts to the original approach.  Now the only differences in plugin\characterize.vim versus the original  are:

1. Testing whether `outs` already has a item, adding `+ `&nbsp;to separate each `charseq`:<br/>`let out = empty(outs) ? '' : '+ '`&nbsp; (and consequently the following `let out` becomes `.=`)
2. `let entities = characterize#html_entities(nr2char(nr))` reflecting the change to autoload\characterize.vim, using `\uxxxx`, `\Uxxxxx`, and `\uxxxx\uxxxx` keys, and
3. When the `outs` list has two items, so is comprised of more than one character, when applicable add `= `&nbsp;, the character (between `<>`), and the HTML5 named character reference(s). (That code could be terser, but structurally it is more in keeping with the preceding code.)

Whether `+ `&nbsp;and `= `&nbsp;are wanted as the separators/indicators of the two code points and the combined HTML5 named character reference(s) respectively is a matter of taste, though the output examples look pretty good IMO:

`∵`<br/>\<∵> 8757, U+2235 BECAUSE, \<C-K>:., \&becaus; \&Because; \&because;
A single character with a digraph and more than one HTML5 named character reference

`👍`<br/>\<👍> 128077, U+1F44D THUMBS UP SIGN, :&zwj;+1:, :&zwj;thumbsup:
Emoji with neither a digraph nor an HTML5 named character reference

`ā`<br/>\<ā> 257, U+0101 LATIN SMALL LETTER A WITH MACRON, \<C-K>a-, \&amacr;
A single character with a digraph and one HTML5 named character reference

`ā`<br/>\<a> 97, \141, U+0061 LATIN SMALL LETTER A + < ̄> 772, U+0304 COMBINING MACRON
A combined character, visually the same as above, with neither a digraph nor an HTML5 named character reference

`∂̸`<br/>\<∂> 8706, U+2202 PARTIAL DIFFERENTIAL, \<C-K>dP, \&part; \&PartialD; + < ̸> 824, U+0338 COMBINING LONG SOLIDUS OVERLAY = <∂̸> \&npart;
A combined character, which is comprised of a character, which itself can be represented by an HTML5 named character reference, plus a combining character, and which together are also an HTML5 named character reference

`	`<br/>\<^I> 9, \011, U+0009 CHARACTER TABULATION, \<C-K>HT, \&Tab;
Control code with digraph and HTML5 named character reference (the only one other than \&NewLine;)

**Comment by kennypete** [comment]
Also updated for Unicode 15, using UnicodeData.txt from https://www.unicode.org/Public/UCD/latest/ucd/UCD.zip

**Comment by tpope** [comment]
Going with a similar approach where we only use `|` for the multi-character case, thanks.

**Pull request closed without merging**